<a href="https://colab.research.google.com/github/paulalcssantos/ClassificacaoKNN/blob/main/1_Classifica%C3%A7%C3%A3o_com_KNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#KNN

*O algoritmo KNN (K-Nearest Neighbors) é um algoritmo de aprendizado de máquina supervisionado usado para classificação e regressão. Ele funciona classificando um novo ponto de dados com base na classe dos seus k vizinhos mais próximos no conjunto de dados de treinamento.*

Será utilizado o famoso Iris Dataset na aplicação do modelo:

https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris

## Importando os dados

In [1]:
from sklearn.datasets import load_iris

##Separando dados entre treino e teste

In [2]:
from sklearn.model_selection import train_test_split

##Importando o KNN

In [3]:
from sklearn.neighbors import KNeighborsClassifier

In [4]:
df = load_iris(as_frame=True)['data']
df['type'] = load_iris(as_frame=True)['target']

df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     type  
0       0  
1       0  
2       0  
3       0  
4       0  
..    ...  
145     2  
146     2  
147     2  
148     2  
149     2  

[150 rows x 5 columns]

In [5]:
df.groupby('type').describe()['sepal length (cm)']

count   mean       std  min    25%  50%  75%  max
type                                                   
0      50.0  5.006  0.352490  4.3  4.800  5.0  5.2  5.8
1      50.0  5.936  0.516171  4.9  5.600  5.9  6.3  7.0
2      50.0  6.588  0.635880  4.9  6.225  6.5  6.9  7.9

In [6]:
df.groupby('type').describe()['petal width (cm)']

count   mean       std  min  25%  50%  75%  max
type                                                 
0      50.0  0.246  0.105386  0.1  0.2  0.2  0.3  0.6
1      50.0  1.326  0.197753  1.0  1.2  1.3  1.5  1.8
2      50.0  2.026  0.274650  1.4  1.8  2.0  2.3  2.5

##Separando os dados

**1. load_iris(as_frame=True):**

    A função load_iris carrega o conjunto de dados Iris.
    O parâmetro as_frame=True indica que os dados devem ser retornados como um pandas DataFrame, o que facilita a manipulação e análise dos dados.

**2. ['data']:**

    O conjunto de dados Iris retornado por load_iris é um objeto do tipo Bunch, que é semelhante a um dicionário.
    A chave 'data' acessa o DataFrame que contém as características (features) das amostras de flores. Estas características são: comprimento da sépala, largura da sépala, comprimento da pétala e largura da pétala.

**3. X = load_iris(as_frame=True)['data']:**

    Aqui, X é atribuído ao DataFrame contendo as características das amostras de flores.
    X terá 150 linhas e 4 colunas, correspondendo às 150 amostras e suas 4 características.

**4.['target']:**

    A chave 'target' acessa a série (ou coluna) que contém os rótulos das classes para cada amostra. Estes rótulos representam os tipos de íris (Iris-setosa, Iris-versicolor, e Iris-virginica), codificados como 0, 1 e 2, respectivamente.

**5. y = load_iris(as_frame=True)['target']:**

    Aqui, y é atribuído à série contendo os rótulos das classes das amostras de flores.
    y terá 150 valores, um para cada amostra.

In [7]:
#Características da pétala
X = load_iris(as_frame=True)['data']

#Tipo de íris
y = load_iris(as_frame=True)['target']

###Treinamento e Teste

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [10]:
#Algoritmo sem dados nenhum, onde K = 3

knn = KNeighborsClassifier(n_neighbors=3)

In [11]:
#Treinando o algoritmo

knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

### Fazendo previsões

In [12]:
#Usando o algoritmo já treinado para fazer previsões usando os dados de teste

y_pred = knn.predict(X_test)

y_pred

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 2, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0, 2, 1, 0, 2, 2, 1, 0, 2])

In [13]:
# Verificando a quantidade de casos errados

print("Número de íris que previmos errado de um total de {0} observações de teste: {1}"\
      .format(X_test.shape[0], (y_test != y_pred).sum()))

Número de íris que previmos errado de um total de 38 observações de teste: 1


##Deixando os dados com a mesma média e mesma variância

In [14]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [16]:
import pandas as pd

In [17]:
pd.DataFrame(X_train).describe()

0             1             2             3
count  1.120000e+02  1.120000e+02  1.120000e+02  1.120000e+02
mean   1.942890e-16  1.003166e-15  8.723181e-17 -1.447255e-16
std    1.004494e+00  1.004494e+00  1.004494e+00  1.004494e+00
min   -1.829120e+00 -2.422105e+00 -1.504166e+00 -1.443850e+00
25%   -9.068399e-01 -5.798248e-01 -1.239194e+00 -1.186757e+00
50%   -9.984503e-02 -1.192548e-01  2.530187e-01  9.870515e-02
75%    7.071499e-01  5.716004e-01  7.829633e-01  8.699826e-01
max    2.321140e+00  3.104736e+00  1.731285e+00  1.641260e+00

In [18]:
#algoritmo sem nenhum dado, onde K = 3

knn = KNeighborsClassifier(n_neighbors=3)

In [19]:
#Treinando o algoritmo

knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [20]:
#Realizando as previsões

y_pred = knn.predict(X_test)

y_pred

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 2, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0, 2, 1, 0, 2, 2, 1, 0, 2])

In [21]:
# Verificando a quantidade de casos errados

print("Número de íris que previmos errado de um total de {0} observações de teste: {1}"\
      .format(X_test.shape[0], (y_test != y_pred).sum()))

Número de íris que previmos errado de um total de 38 observações de teste: 1


##Calculando o melhor valor de K

In [22]:
percentual_acerto = []
tamanho_teste = X_test.shape[0]

for i in range(1, 15):
  knn = KNeighborsClassifier(n_neighbors=i)
  knn.fit(X_train, y_train)
  pred_i = knn.predict(X_test)

  acerto_i = (y_test == pred_i).sum()/tamanho_teste
  percentual_acerto.append(100 * acerto_i)

In [23]:
import pandas as pd

In [24]:
df_acerto = pd.DataFrame()
df_acerto['K'] = range(1, 15)
df_acerto['percetual_acerto'] = percentual_acerto

df_acerto.head()

K  percetual_acerto
0  1         92.105263
1  2         94.736842
2  3         97.368421
3  4         97.368421
4  5         97.368421

##Gerando gráfico

In [25]:
import plotly.express as px

In [28]:
fig = px.line(df_acerto, x='K', y='percetual_acerto',
              title="Percentual de Acerto por K")
fig.show()